### PR TITLE
Include hash of program information in metadata

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

Add a SHA256 hash of the program information used to verify that a program is safe to the program metadata. A subsequent change in EC will verify that the hash matches prior to permitting the attach to complete.

## Testing

CI/Cd

## Documentation

No.

### CODE REVIEW

File: include/ebpf_api.h [hunk 1]
```diff
@@ -413,6 +413,10 @@ extern "C"
     ebpf_get_next_pinned_program_path(
         _In_z_ const char* start_path, _Out_writes_z_(EBPF_MAX_PIN_PATH_LENGTH) char* next_path);
 
+    typedef struct _ebpf_program_info ebpf_program_info_t;
+    ebpf_result_t
+    ebpf_get_program_info(_Out_ const ebpf_program_info_t** program_info);
```
Comment (dthaler) [inline include/ebpf_api.h:418]:
```suggestion
    ebpf_get_program_info(_Outptr_ const ebpf_program_info_t** program_info);
```

File: libs/api/ebpf_api.cpp [hunk 2]
```diff
@@ -3256,6 +3256,17 @@ ebpf_get_program_type_by_name(
     EBPF_RETURN_RESULT(result);
 }
 
+ebpf_result_t
+ebpf_get_program_info(_Out_ const ebpf_program_info_t** program_info)
```
Comment (dthaler) [inline libs/api/ebpf_api.cpp:3260]:
```suggestion
ebpf_get_program_info(_Outptr_ const ebpf_program_info_t** program_info)
```

**Review by dthaler — DISMISSED** [review]

File: include/ebpf_api.h [hunk 3]
```diff
@@ -413,6 +413,20 @@ extern "C"
...
+    /**
+     * @brief Get the set of program information used by the verifier during
+     * the last verification.
+     *
+     * @param[out] program_info Pointer to the program information used to
+     * verify the program.
+     * @retval EBPF_SUCCESS The operation was successful.
+     * @retval EBPF_OBJECT_NOT_FOUND No program information was found.
+     */
+    ebpf_result_t
```
Comment (shankarseal) [inline include/ebpf_api.h:427]: suggest change the name to reflect the intent of the API.

**Review by ryihan — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

File: tools/bpf2c/bpf_code_generator.h [hunk 4]
```diff
@@ -302,9 +307,13 @@ class bpf_code_generator
      *
      * @param[in] program_type Program type GUID.
      * @param[in] attach_type Attach type GUID.
+     * @param[in] attach_type Hash of the program information used to verify this program.
```
Comment (shankarseal) [inline tools/bpf2c/bpf_code_generator.h:310]: typo: `program_info_hash`

File: tools/bpf2c/bpf_code_generator.h [hunk 5]
```diff
@@ -220,7 +220,11 @@ class bpf_code_generator
      * @param[in] attach_type Expected attach type GUID for the section.
      */
```
Comment (shankarseal) [inline tools/bpf2c/bpf_code_generator.h:221]: update parameter list.

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]

**Pull request merged**